### PR TITLE
[dhcp_server] Add dhcpservd to dhcp_server container

### PR DESCRIPTION
**Pull request by yaqiangz**
**Depends on: https://github.com/sonic-net/sonic-buildimage/pull/16548**

#### Why I did it
Part implementation of dhcp_server. HLD: https://github.com/sonic-net/SONiC/pull/1282.
Add dhcpservd to dhcp_server container.

##### Work item tracking
- Microsoft ADO **(number only)**: 25093520

#### How I did it
1. Add installing required pkg (psutil) in `Dockerfile`.
2. Add copying required file to container in `Dockerfile` (kea-dhcp related and dhcpservd related)
3. Add critical_process and supervisor config.
4. Add support for generating kea config (only in dhcpservd.py) and updating lease table (in dhcpservd.py and lease_update.sh)

#### How to verify it
1. Build image with setting `INCLUDE_DHCP_SERVER` to y and enabled dhcp_server feature after installed image, container start as expected.
2. Enter container and found that all processes defined in supervisor configuration running as expected.
3. Kill processes defined in critical_processes, container exist.

unit test
```
========================================== test session starts ==========================================
platform linux -- Python 3.9.2, pytest-6.0.2, py-1.10.0, pluggy-0.13.0
rootdir: /sonic/src/sonic-dhcpservd, configfile: setup.cfg, testpaths: tests
plugins: pyfakefs-5.2.4, cov-2.10.1
collected 26 items                                                                                      

tests/test_dhcp_cfggen.py .........                                                               [ 34%]
tests/test_dhcp_lease.py .....                                                                    [ 53%]
tests/test_dhcp_server_utils.py ........                                                          [ 84%]
tests/test_dhcpservd.py ....                                                                      [100%]Clearing the cache


----------- coverage: platform linux, python 3.9.2-final-0 -----------
Name                             Stmts   Miss Branch BrPart     Cover   Missing
-------------------------------------------------------------------------------
dhcpservd/dhcp_cfggen.py           170     16     72      1    92.98%   49-66, 149-153, 239->241, 241
dhcpservd/dhcp_server_utils.py      37      3     14      0    94.12%   35, 55-56
dhcpservd/dhcp_lease.py             96      3     30      4    94.44%   59->61, 61, 62->63, 63, 75->77, 113->135, 154
dhcpservd/dhcpservd.py              31      1      4      0    97.14%   49
-------------------------------------------------------------------------------
TOTAL                              334     23    120      5    93.83%

Required test coverage of 80.0% reached. Total coverage: 93.83%

========================================== slowest 5 durations ==========================================

```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: src/sonic-dhcpservd/dhcpservd.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+KEA_DHCP4_CONFIG = "/etc/kea/kea-dhcp4.conf"
+KEA_DHCP4_PROC_NAME = "kea-dhcp4"
+
+
+class DhcpServd(object):
+    def __init__(self, dhcp_cfg, db_connector):
+        self.dhcp_cfg = dhcp_cfg
+        self.db_connector = db_connector
+
+    def notify_kea_dhcp4_proc(self):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:25]: `apply_config`

File: src/sonic-dhcpservd/dhcpservd.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        Send SIGHUP signal to kea-dhcp4 process
+        """
+        for proc in psutil.process_iter():
+            if KEA_DHCP4_PROC_NAME in proc.name():
+                proc.send_signal(signal.SIGHUP)
+                break
+
+    def generate_kea_dhcp4_config(self, from_db=True, config_file_path=""):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:34]: `generate_dhcp4_config`

File: src/sonic-dhcpservd/dhcpservd.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+KEA_DHCP4_CONFIG = "/etc/kea/kea-dhcp4.conf"
+KEA_DHCP4_PROC_NAME = "kea-dhcp4"
+
+
+class DhcpServd(object):
+    def __init__(self, dhcp_cfg, db_connector):
+        self.dhcp_cfg = dhcp_cfg
+        self.db_connector = db_connector
+
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:24]: `def start(self)`

1. subscribe config db change (tables)
2. dispatch (change tables) -> generate_dhcp4_config() or generate_dhcp6_config() or both
3. apply_config

File: src/sonic-dhcpservd/dhcpservd.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    INIT_CONFIG_FILE
+from dhcp_server_utils import DhcpDbConnector
+from functools import partial
+
+KEA_DHCP4_CONFIG = "/etc/kea/kea-dhcp4.conf"
+KEA_DHCP4_PROC_NAME = "kea-dhcp4"
+
+
+class DhcpServd(object):
+    def __init__(self, dhcp_cfg, db_connector):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:21]: dhcp_cfg_generator

File: src/sonic-dhcpservd/dhcpservd.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                break
+
+    def generate_kea_dhcp4_config(self, from_db=True, config_file_path=""):
+        """
+        Generate kea-dhcp4 config file and dump it to config folder
+        Args:
+            from_db: boolean, if set to True, generate config from running config_db
+            config_file_path: str, if from_db is False, generate config from config_db file
+        """
+        kea_dhcp4_config = self.dhcp_cfg.generate_kea_dhcp4_config(from_db, config_file_path)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:41]: Use exception handling errors

File: src/sonic-dhcpservd/dhcpservd.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        if kea_dhcp4_config is None:
+            syslog.syslog(syslog.LOG_ERR, "Cannot get kea-dhcp4 configure")
+            return
+        try:
+            with open(KEA_DHCP4_CONFIG, "w") as write_file:
+                json.dump(kea_dhcp4_config, write_file, indent=4, ensure_ascii=False)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:47]: unicode

File: src/sonic-dhcpservd/dhcpservd.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                # After refresh kea-config, we need to SIGHUP kea-dhcp4 process to read new config
+                self.notify_kea_dhcp4_proc()
+        except FileNotFoundError:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:50]: Do not capture exceptions that you don't know how to handle it

File: src/sonic-dhcpservd/dhcpservd.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+            syslog.syslog(syslog.LOG_ERR, "Cannot write to: {}".format(INIT_CONFIG_FILE))
+
+    def update_lease_handler(self, signum, frame):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd.py:53]: `class LeaseManager(object)`
- list[] = LeaseHandler objects
- `start(self`) -> for obj in list: obj.register()
- `stop`

`class LeaseHandler(object)`
- `read(self)` abstract method
- `register(self)` abstract method
- `unregister`
- `update(self)` for lease in self.read() ... write state_db ...

`class KeaDhcp4LeaseHandler(LeaseHandler)`
- `read(self)` return list of leases
- `register(self)` -> `signal.signal(signal.SIGUSR1, self.update)`

File: src/sonic-dhcpservd/dhcp_cfggen.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+DHCP_SERVER_IPV4 = "DHCP_SERVER_IPV4"
+DHCP_SERVER_IPV4_CUSTOMIZED_OPTIONS = "DHCP_SERVER_IPV4_CUSTOMIZED_OPTIONS"
+DHCP_SERVER_IPV4_RANGE = "DHCP_SERVER_IPV4_RANGE"
+DHCP_SERVER_IPV4_PORT = "DHCP_SERVER_IPV4_PORT"
+DHCP_SERVER_IPV4_LEASE = "DHCP_SERVER_IPV4_LEASE"
+# Default lease time of DHCP
+DEFAULT_LEASE_TIME = 900
+
+
+class DhcpServCfg(object):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcp_cfggen.py:22]: DhcpCfgGenerator

File: src/sonic-dhcpservd/dhcp_cfggen.py [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                self.kea_config = json.load(fp)
+        except FileNotFoundError:
+            syslog.syslog(syslog.LOG_ERR, "Cannot find init config file {}".format(INIT_CONFIG_FILE))
+            return None
+        except json.decoder.JSONDecodeError:
+            syslog.syslog(syslog.LOG_ERR, "Incorrect format of {}".format(INIT_CONFIG_FILE))
+            return None
+
+        # Generate from running config_db
+        if from_db:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcp_cfggen.py:291]: `class DhcpConfigProvider(object)`
- `def get_dhcp_config(self)`
 return {
   'device_metadata': self.get_device_metadta()
   'hostname': self.get_dhcp_config()
 }

`class RunningConfigDchpConfigProvider(DhcpConfigProvider)`
- init - db_connector

`class FileBasedDchpConfigProvider(DhcpConfigProvider)`
- init - cfg_file_path

db_connector = DbConnector(...)
config_provider = RunningConfigDchpConfigProvider(db_connector)
cfggen1 = DhcpConfigGenerator(config_provider)

config_provider = FileBasedDchpConfigProvider("/etc/sonic/config_db.json")
cfggen2 = DhcpConfigGenerator(config_provider)

File: src/sonic-dhcpservd/dhcp_cfggen.py [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+    def generate_kea_dhcp4_config(self, from_db=True, config_file_path=""):
+        """
+        Generate kea-dhcp4 config
+        Args:
+            from_db: boolean, if set to True, generate config from running config_db
+            config_file_path: str, if from_db is False, generate config from config_db file
+        Returns:
+            config dict
+        """
+        try:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcp_cfggen.py:280]: lazy load
if self.kea_config == None
   self.kea_config = json.load...

output = deepcopy(self.kea_config)
...
return output

File: src/sonic-dhcpservd/dhcp_cfggen.py [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+        {
+            'range2': [IPv4Address('192.168.0.3'), IPv4Address('192.168.0.6')],
+            'range1': [IPv4Address('192.168.0.2'), IPv4Address('192.168.0.5')],
+            'range3': [IPv4Address('192.168.0.10'), IPv4Address('192.168.0.10')]
+        }
+        Args:
+            range_ipv4: Table object or dict of range.
+        """
+        self.ranges = {}
+        is_dict, keys = get_keys(range_ipv4)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcp_cfggen.py:118]: `get_values`?

File: dockers/docker-dhcp-server/kea-dhcp4.conf.j2 [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+{%- if dhcp_server_ipv4 %}
+    {%- for dhcp_server_entry in dhcp_server_ipv4 if "state" in dhcp_server_ipv4[dhcp_server_entry] and dhcp_server_ipv4[dhcp_server_entry]["state"] == "enabled" and dhcp_server_entry in vlan_interface %}
+        {%- if "mode" in dhcp_server_ipv4[dhcp_server_entry] %}
+            {%- if dhcp_server_ipv4[dhcp_server_entry]["mode"] == "PORT" %}
+                {%- for dhcp_interface_ip in port_ips[dhcp_server_entry] %}
+            {
+                "subnet": "{{ port_ips[dhcp_server_entry][dhcp_interface_ip]["subnet"] }}",
+                "pools": [
+                {%- set server_id = dhcp_interface_ip.split("/")[0] -%}
+                {%- set lease_time = default_lease_time %}
```
Comment (Blueve) [inline dockers/docker-dhcp-server/kea-dhcp4.conf.j2:38]: define this in j2 template

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+# Default lease time of DHCP
+DEFAULT_LEASE_TIME = 900
+
+
+class DhcpServCfgGenerator(object):
+    def __init__(self, dhcp_db_connector):
+        # Read port alias map file, this file is render after container start, so it would not change any more
+        self.port_alias_map = {}
+        self.db_connector = dhcp_db_connector
+        with open(PORT_MAP_PATH, "r") as file:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:31]: Move to constructor
PORT_MAP_PATH
KEA_TEAMPLTE_PATH

os.path

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+                    str(ipaddress.ip_network(dhcp_interface_ip, strict=False))
+
+        # Store network - dhcp port map file which would be used by dhcpservd while updating lease table
+        ip_ports = {}
+        for key, value in self.ip_ports.items():
+            ip_ports[str(key)] = value
+        with open(DHCP_SERVER_IP_PORTS_FILE, "w") as write_file:
+            json.dump(ip_ports, write_file, indent=4, ensure_ascii=False)
+
+    def generate_kea_dhcp4_config(self):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:246]: Move public method(API) to front

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+            config_file_path: str, if from_db is False, generate config from config_db file
+        Returns:
+            config dict
+        """
+        # Generate from running config_db
+        # Get host name
+        device_metadata = self.db_connector.get_config_db_table("DEVICE_METADATA")
+        localhost_entry = device_metadata.get("localhost", {})
+        if localhost_entry is None or "hostname" not in localhost_entry:
+            syslog.syslog(syslog.LOG_ERR, "Cannot get hostname")
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:260]: raise exception

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+        # Get ip information of vlan
+        vlan_interface = self.db_connector.get_config_db_table("VLAN_INTERFACE")
+        vlan_interface_keys = set(vlan_interface.keys())
+        vlan_interfaces = self._get_vlan_ipv4_interface(vlan_interface_keys)
+        dhcp_server_ipv4, customized_options_ipv4, range_ipv4, port_ipv4 = self._get_dhcp_ipv4_tables_from_db()
+        vlan_member_table = self.db_connector.get_config_db_table("VLAN_MEMBER")
+        vlan_members = set(vlan_member_table.keys())
+
+        # Parse range table
+        self._parse_range(range_ipv4)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:272]: ranges = self._parse_range(..)
options = self.parse_options(.., ranges)

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+            "dhcp_server_ipv4": dhcp_server_ipv4,
+            "vlan_interface": vlan_interfaces,
+            "port_ips": self.port_ips,
+            "default_lease_time": DEFAULT_LEASE_TIME,
+            "hostname": hostname,
+            "lease_path": DEFAULT_LEASE_PATH,
+            "lease_update_script_path": LEASE_UPDATE_SCRIPT_PATH
+        }
+        output = self.kea_template.render(j2_obj)
+        return output
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:295]: 1. parse xxx from dbconnector
2. generate kea config
3. render config

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+        Register lease hanlder
+        """
+        for handler in self.lease_handlers:
+            handler.register()
+
+
+class LeaseHanlder(object):
+    def __init__(self, db_connector):
+        self.db_connector = db_connector
+        self.lease_update_interval = DEFAULE_LEASE_UPDATE_INTERVAL
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:31]: move to constructor

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+        last_update_time = self.last_update_time
+        curr_time = datetime.now()
+        # If the time since the last update is less than self.lease_update_interval, then wait for a
+        # self.lease_update_interval
+        if last_update_time is not None and (curr_time - last_update_time).seconds < self.lease_update_interval:
+            time.sleep(self.lease_update_interval)
+            if self.last_update_time != last_update_time:
+                # Means lease has been updated during sleep, no need to update this lease
+                return
+        new_lease = self._read()
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:66]: acquire lock before refresh leases

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+                if key in old_lease_key:
+                    self.db_connector.state_db.delete("{}|{}".format(DHCP_SERVER_IPV4_LEASE, key))
+                continue
+            new_key = "{}|{}".format(DHCP_SERVER_IPV4_LEASE, key)
+            for k, v in new_lease[key].items():
+                self.db_connector.state_db.hset(new_key, k, v)
+
+        # Delete old lease not in new lease set
+        for key in old_lease_key:
+            if key not in new_lease_key:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:88]: key not in new_lease.keys()

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+
+    @abstractmethod
+    def _read(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def register(self):
+        raise NotImplementedError
+
+    def _update_lease(self, signum, frame):
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:42]: def update_leases(self):

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+                        "ip": ip_str
+                    }
+                    break
+        return new_lease
+
+    def register(self):
+        """
+        Register callback function of signal
+        """
+        signal.signal(signal.SIGUSR1, self._update_lease)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:158]: def _update_lease(self, signum, frame):
    self.update_lease()

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+        self.last_update_time = datetime.now()
+
+
+class KeaDhcp4LeaseHandler(LeaseHanlder):
+    def __init__(self, db_connector, lease_file=KEA_LEASE_FILE_PATH):
+        LeaseHanlder.__init__(self, db_connector)
+        self.lease_file = lease_file
+
+    def _read(self):
+        """
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:100]: Move to base method definition

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+        self.lease_update_interval = DEFAULE_LEASE_UPDATE_INTERVAL
+        self.last_update_time = None
+
+    @abstractmethod
+    def _read(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def register(self):
+        raise NotImplementedError
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:40]: doc comment

File: src/sonic-dhcpservd/dhcpservd/dhcpservd.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+            self._notify_kea_dhcp4_proc()
+
+    def start(self):
+        self.dump_dhcp4_config()
+        lease_manager = LeaseManager(self.db_connector, KEA_LEASE_FILE_PATH)
+        lease_manager.start()
+
+        # TODO Add config db subcribe to re-generate kea-dhcp4 config after config_db change.
+
+        while True:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcpservd.py:55]: make `start` non-block

File: src/sonic-dhcpservd/dhcpservd/dhcpservd.py [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+
+    def dump_dhcp4_config(self):
+        """
+        Generate kea-dhcp4 config file and dump it to config folder
+        Args:
+            from_db: boolean, if set to True, generate config from running config_db
+            config_file_path: str, if from_db is False, generate config from config_db file
+        """
+        kea_dhcp4_config = self.dhcp_cfg.generate()
+        if kea_dhcp4_config is None:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcpservd.py:40]: remove this line

File: src/sonic-dhcpservd/tests/test_dhcp_server_utils.py [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+            call(ANY, 'key1'),
+            call().items(),
+            call().items().__iter__(),
+            call(ANY, 'key2'),
+            call().items(),
+            call().items().__iter__()
+        ])
+
+
+def test_merge_intervals():
```
Comment (Blueve) [inline src/sonic-dhcpservd/tests/test_dhcp_server_utils.py:43]: add more cases

File: src/sonic-dhcpservd/tests/test_dhcp_cfggen.py [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+                                                  port_map_path="tests/test_data/port-name-alias-map.txt",
+                                                  kea_conf_template_path="tests/test_data/kea-dhcp4.conf.j2")
+        config = dhcp_cfg_generator.generate()
+        # Verify whether configuration generated as expected.
+        if test_config_db == "mock_config_db.json":
+            assert json.loads(config) == expected_dhcp_config
+            # While generating configuration, we will use json.dump to save net-interface information, verify that.
+            mock_json_dump.assert_called_once_with({"192.168.0.0/21": "Vlan1000"}, ANY, indent=4, ensure_ascii=False)
+        else:
+            assert json.loads(config) == expected_dhcp_config_without_port_config
```
Comment (Blueve) [inline src/sonic-dhcpservd/tests/test_dhcp_cfggen.py:152]: make field orderding not impact object comparison result

File: src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py [hunk 30]
```diff
@@ -0,0 +1,334 @@
...
+                    # Loop the IP of the dhcp interface and find the network that target range is in this network.
+                    self._match_range_network(dhcp_interface, dhcp_interface_name, port, range, port_ips)
+        # Merge ranges to avoid overlap
+        for dhcp_interface_name, value in port_ips.items():
+            for dhcp_interface_ip, port_range in value.items():
+                for port_name, ip_range in port_range.items():
+                    ranges = merge_intervals(ip_range)
+                    ranges = [[str(range[0]), str(range[1])] for range in ranges]
+                    port_ips[dhcp_interface_name][dhcp_interface_ip][port_name] = ranges
+        # Store network - dhcp port map file which would be used by dhcpservd while updating lease table
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_cfggen.py:331]: Remove this dump

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+        # Read lease file generated by kea-dhcp4
+        try:
+            with open(self.lease_file, "r", encoding="utf-8") as fb:
+                dq = deque(fb)
+        except FileNotFoundError as err:
+            syslog.syslog(syslog.LOG_ERR, "Cannot find lease file: {}".format(self.lease_file))
+            raise err
+
+        # Read network - dhcp interface map file, this file is generated by dhcp_cfggen while generate config
+        with open(self.ip_ports_file, "r", encoding="utf8")as fp:
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:113]: Build this mapping from config db directly

File: src/sonic-dhcpservd/dhcpservd/dhcp_lease.py [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+            if splits[0] == "address":
+                break
+            ip_str = splits[0]
+            mac_address = splits[1]
+            valid_lifetime = splits[3]
+            lease_end = splits[4]
+
+            for net in nets:
+                if ipaddress.ip_address(ip_str) in net:
+                    new_key = "{}|{}".format(net_intf_map[net], mac_address)
```
Comment (Blueve) [inline src/sonic-dhcpservd/dhcpservd/dhcp_lease.py:135]: What happens if two vlan configured same subnet?
VLAN1000 - 192.168.0.1/24
VLAN2000 - 192.168.0.1/24

Current idea:
lease <--> Assigned IP --belong to ?---> VLAN?

How about use IP address?
lease <--> client MAC ----FBD----> VLAN?

File: rules/config [hunk 33]
```diff
@@ -151,7 +151,7 @@ INCLUDE_NAT = y
 INCLUDE_DHCP_RELAY = y
 
 # INCLUDE_DHCP_SERVER - build and install dhcp-server package
-INCLUDE_DHCP_SERVER = n
+INCLUDE_DHCP_SERVER = y
```
Comment (Blueve) [inline rules/config:154]: Change to 'n' when you ready for reviewing

**Review by Blueve — APPROVED** [review]

**Comment by yaqiangz** [comment]
@yxieca Could you pls help to merge this PR?

**Pull request merged**